From 6bfe1394d6db7c26e2f1beddcfb77f04a12ead87 Mon Sep 17 00:00:00 2001
From: Florent Couzon <prplmesh.opensources@sagemcom.com>
Date: Mon, 10 Feb 2025 11:53:42 +0100
Subject: [PATCH] slave: fill actions upon m8 reception

Current situation is that M8 handling triggers placeholders functions in agent.

This commit adds handling for enabling disabling EP (the function in sta_wlan_hal_whm is to be develloped).
It also adds handling for bsta_reconfiguration from ap_autoconfiguration to backhaul manager. The backhaul manager handling comes in a future commit.

Signed-off-by: Florent Couzon <prplmesh.opensources@sagemcom.com>
---
 .../backhaul_manager/backhaul_manager.cpp     | 37 +++++++++
 .../slave/backhaul_manager/backhaul_manager.h |  1 +
 .../slave/tasks/ap_autoconfiguration_task.cpp | 66 +++++++++++++--
 .../slave/tasks/ap_autoconfiguration_task.h   |  3 +-
 .../beerocks/bwl/dummy/sta_wlan_hal_dummy.cpp |  2 +
 .../beerocks/bwl/dummy/sta_wlan_hal_dummy.h   |  2 +
 .../beerocks/bwl/dwpal/sta_wlan_hal_dwpal.cpp |  2 +
 .../beerocks/bwl/dwpal/sta_wlan_hal_dwpal.h   |  2 +
 .../bwl/dwpald/sta_wlan_hal_dwpald.cpp        |  2 +
 .../beerocks/bwl/dwpald/sta_wlan_hal_dwpald.h |  2 +
 .../beerocks/bwl/include/bwl/sta_wlan_hal.h   |  2 +
 .../bwl/nl80211/sta_wlan_hal_nl80211.cpp      |  2 +
 .../bwl/nl80211/sta_wlan_hal_nl80211.h        |  2 +
 common/beerocks/bwl/whm/sta_wlan_hal_whm.cpp  |  6 ++
 common/beerocks/bwl/whm/sta_wlan_hal_whm.h    |  2 +
 .../beerocks/tlvf/beerocks_message_action.h   |  3 +-
 .../beerocks/tlvf/beerocks_message_backhaul.h | 23 ++++++
 .../tlvf/beerocks_message_backhaul.cpp        | 80 +++++++++++++++++++
 .../tlvf/beerocks_message_action.yaml         |  3 +-
 .../tlvf/beerocks_message_backhaul.yaml       |  6 ++
 20 files changed, 238 insertions(+), 10 deletions(-)

diff --git a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
index 15e1f7130..75326aa7a 100644
--- a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
+++ b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
@@ -1812,6 +1812,28 @@ bool BackhaulManager::handle_slave_backhaul_message(int fd, ieee1905_1::CmduMess
         }
         break;
     }
+    case beerocks_message::ACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT: {
+        LOG(TRACE) << "received ACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT";
+        auto request =
+            beerocks_header
+                ->addClass<beerocks_message::cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT>();
+        if (!request) {
+            LOG(ERROR) << "addClass ACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT failed";
+            return false;
+        }
+
+        const sMacAddr &radio_mac = request->radio_mac();
+
+        LOG(DEBUG) << "handle ACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT for bsta " << radio_mac;
+        auto sta_wlan_hal = get_wireless_hal(radio_mac);
+        if (!sta_wlan_hal) {
+            return false;
+        }
+
+        sta_wlan_hal->enable_disable_ep(static_cast<bool>(request->enable()));
+
+        break;
+    }
     default: {
         bool handled =
             m_task_pool.handle_cmdu(cmdu_rx, 0, sMacAddr(), sMacAddr(), fd, beerocks_header);
@@ -2603,6 +2625,21 @@ std::shared_ptr<bwl::sta_wlan_hal> BackhaulManager::get_wireless_hal(std::string
     return {};
 }
 
+std::shared_ptr<bwl::sta_wlan_hal> BackhaulManager::get_wireless_hal(const sMacAddr &radio_mac)
+{
+    auto radio_it = std::find_if(m_radios_info.begin(), m_radios_info.end(),
+                                 [&radio_mac](const std::shared_ptr<sRadioInfo> radio_info) {
+                                     return radio_info->radio_mac == radio_mac;
+                                 });
+
+    if (radio_it == m_radios_info.end()) {
+        LOG(ERROR) << "Invalid radio " << radio_mac;
+        return nullptr;
+    }
+
+    return (*radio_it)->sta_wlan_hal;
+}
+
 bool BackhaulManager::handle_slave_failed_connection_message(ieee1905_1::CmduMessageRx &cmdu_rx,
                                                              const sMacAddr &src_mac)
 {
diff --git a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.h b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.h
index e8b783cf8..c19c22d6a 100644
--- a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.h
+++ b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.h
@@ -251,6 +251,7 @@ private:
     clear_radio_handlers(const std::shared_ptr<beerocks::BackhaulManager::sRadioInfo> &radio_info);
 
     std::shared_ptr<bwl::sta_wlan_hal> get_wireless_hal(std::string iface = "");
+    std::shared_ptr<bwl::sta_wlan_hal> get_wireless_hal(const sMacAddr &radio_mac);
 
 private:
     /**
diff --git a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
index 53586965f..6d1927258 100755
--- a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
+++ b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
@@ -1293,7 +1293,7 @@ void ApAutoConfigurationTask::handle_ap_autoconfiguration_wsc(ieee1905_1::CmduMe
                 send_bsta_configuration(radio->front.iface_mac, *bSTA_it);
                 configs.erase(bSTA_it);
             } else if (db->controller_info.early_ap_capability) {
-                send_enable_disable_endpoint(radio->front.iface_mac, false, true);
+                send_enable_disable_endpoint(radio->front.iface_mac, false);
             }
             send_ap_bss_configuration_message(radio->front.iface_name, configs);
         } else {
@@ -1957,15 +1957,69 @@ bool ApAutoConfigurationTask::handle_agent_ap_mld_configuration_tlv(
 bool ApAutoConfigurationTask::send_bsta_configuration(const sMacAddr &radio_mac,
                                                       const WSC::configData::config &config)
 {
-    // Place holder
-    return true;
+    auto request = message_com::create_vs_message<
+        beerocks_message::cACTION_BACKHAUL_WIFI_CREDENTIALS_UPDATE_REQUEST>(m_cmdu_tx);
+    if (!request) {
+        LOG(ERROR) << "Failed building message ACTION_BACKHAUL_WIFI_CREDENTIALS_UPDATE_REQUEST!";
+        return false;
+    }
+    auto backhaul_manager_cmdu_client = m_btl_ctx.get_backhaul_manager_cmdu_client();
+    if (!backhaul_manager_cmdu_client) {
+        LOG(ERROR) << "Failed to get backhaul manager cmdu client";
+        return false;
+    }
+
+    request->radio_mac() = radio_mac;
+
+    std::stringstream ss;
+    auto bSta_credentials = request->create_wifi_credentials();
+    if (!bSta_credentials) {
+        LOG(ERROR) << "Failed building message!";
+        return false;
+    }
+
+    ss << "bSTA: " << config.bssid << std::endl;
+
+    bSta_credentials->bssid_attr().data = config.bssid;
+    bSta_credentials->bss_type()        = config.bss_type;
+
+    ss << "- SSID: " << config.ssid << std::endl
+       << "- Key: " << config.network_key << std::endl
+       << "- Auth: " << std::hex << int(config.auth_type) << std::endl
+       << "- Encr: " << std::hex << int(config.encr_type) << std::endl;
+
+    bSta_credentials->set_ssid(config.ssid);
+    bSta_credentials->set_network_key(config.network_key);
+    bSta_credentials->authentication_type_attr().data = config.auth_type;
+    bSta_credentials->encryption_type_attr().data     = config.encr_type;
+    request->add_wifi_credentials(bSta_credentials);
+
+    LOG(INFO) << "Sending bSTA configuration: " << std::endl << ss.str();
+
+    LOG(INFO) << "Sending ACTION_BACKHAUL_WIFI_CREDENTIALS_UPDATE_REQUEST to BH manager";
+    return backhaul_manager_cmdu_client->send_cmdu(m_cmdu_tx);
 }
 
 bool ApAutoConfigurationTask::send_enable_disable_endpoint(const sMacAddr &radio_mac,
-                                                           const bool enable, const bool force)
+                                                           const bool enable)
 {
-    // Place holder
-    return true;
+    auto request = message_com::create_vs_message<
+        beerocks_message::cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT>(m_cmdu_tx);
+    if (!request) {
+        LOG(ERROR) << "Failed building message ACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT!";
+        return false;
+    }
+    auto backhaul_manager_cmdu_client = m_btl_ctx.get_backhaul_manager_cmdu_client();
+    if (!backhaul_manager_cmdu_client) {
+        LOG(ERROR) << "Failed to get backhaul manager cmdu client";
+        return false;
+    }
+
+    request->radio_mac() = radio_mac;
+    request->enable()    = enable;
+
+    LOG(INFO) << "Sending ACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT to BH manager";
+    return backhaul_manager_cmdu_client->send_cmdu(m_cmdu_tx);
 }
 
 void ApAutoConfigurationTask::handle_vs_wifi_credentials_update_response(
diff --git a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.h b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.h
index 29ee13945..4ea78b857 100644
--- a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.h
+++ b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.h
@@ -227,8 +227,7 @@ private:
 
     bool send_bsta_configuration(const sMacAddr &radio_mac, const WSC::configData::config &config);
 
-    bool send_enable_disable_endpoint(const sMacAddr &radio_mac, const bool enable,
-                                      const bool force);
+    bool send_enable_disable_endpoint(const sMacAddr &radio_mac, const bool enable);
 
     bool validate_reconfiguration(const std::string &radio_iface,
                                   std::vector<WSC::configData::config> &configs);
diff --git a/common/beerocks/bwl/dummy/sta_wlan_hal_dummy.cpp b/common/beerocks/bwl/dummy/sta_wlan_hal_dummy.cpp
index b8ea61cf2..597558e1b 100644
--- a/common/beerocks/bwl/dummy/sta_wlan_hal_dummy.cpp
+++ b/common/beerocks/bwl/dummy/sta_wlan_hal_dummy.cpp
@@ -91,6 +91,8 @@ std::string sta_wlan_hal_dummy::get_wireless_backhaul_mac()
     return mac;
 }
 
+bool sta_wlan_hal_dummy::enable_disable_ep(bool enable) { return true; }
+
 bool sta_wlan_hal_dummy::process_dummy_data(parsed_obj_map_t &parsed_obj) { return true; }
 
 bool sta_wlan_hal_dummy::process_dummy_event(parsed_obj_map_t &parsed_obj) { return true; }
diff --git a/common/beerocks/bwl/dummy/sta_wlan_hal_dummy.h b/common/beerocks/bwl/dummy/sta_wlan_hal_dummy.h
index dca6fb15d..f98a066e6 100644
--- a/common/beerocks/bwl/dummy/sta_wlan_hal_dummy.h
+++ b/common/beerocks/bwl/dummy/sta_wlan_hal_dummy.h
@@ -66,6 +66,8 @@ public:
     std::string get_bssid() override;
     std::string get_wireless_backhaul_mac() override;
 
+    virtual bool enable_disable_ep(bool enable) override;
+
 protected:
     virtual bool process_dummy_data(parsed_obj_map_t &parsed_obj) override;
     virtual bool process_dummy_event(parsed_obj_map_t &parsed_obj) override;
diff --git a/common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.cpp b/common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.cpp
index bfdaf618a..794b6a31a 100644
--- a/common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.cpp
+++ b/common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.cpp
@@ -476,6 +476,8 @@ std::string sta_wlan_hal_dwpal::get_wireless_backhaul_mac()
     return mac;
 }
 
+bool sta_wlan_hal_dwpal::enable_disable_ep(bool enable) { return true; }
+
 bool sta_wlan_hal_dwpal::process_dwpal_event(char *buffer, int bufLen, const std::string &opcode)
 {
     LOG(TRACE) << __func__ << " " << get_iface_name() << " - opcode: |" << opcode << "|";
diff --git a/common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.h b/common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.h
index 6468f8a11..40dc99371 100644
--- a/common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.h
+++ b/common/beerocks/bwl/dwpal/sta_wlan_hal_dwpal.h
@@ -63,6 +63,8 @@ public:
     std::string get_bssid() override;
     std::string get_wireless_backhaul_mac() override;
 
+    virtual bool enable_disable_ep(bool enable) override;
+
 protected:
     virtual bool process_dwpal_event(char *buffer, int bufLen, const std::string &opcode) override;
     virtual bool process_dwpal_nl_event(struct nl_msg *msg, void *arg = nullptr) override;
diff --git a/common/beerocks/bwl/dwpald/sta_wlan_hal_dwpald.cpp b/common/beerocks/bwl/dwpald/sta_wlan_hal_dwpald.cpp
index d0334a718..fdac07226 100644
--- a/common/beerocks/bwl/dwpald/sta_wlan_hal_dwpald.cpp
+++ b/common/beerocks/bwl/dwpald/sta_wlan_hal_dwpald.cpp
@@ -502,6 +502,8 @@ std::string sta_wlan_hal_dwpal::get_wireless_backhaul_mac()
     return mac;
 }
 
+bool sta_wlan_hal_dwpal::enable_disable_ep(bool enable) { return true; }
+
 bool sta_wlan_hal_dwpal::process_dwpal_event(char *ifname, char *buffer, int bufLen,
                                              const std::string &opcode)
 {
diff --git a/common/beerocks/bwl/dwpald/sta_wlan_hal_dwpald.h b/common/beerocks/bwl/dwpald/sta_wlan_hal_dwpald.h
index 124014feb..5e6ba0971 100644
--- a/common/beerocks/bwl/dwpald/sta_wlan_hal_dwpald.h
+++ b/common/beerocks/bwl/dwpald/sta_wlan_hal_dwpald.h
@@ -63,6 +63,8 @@ public:
     std::string get_bssid() override;
     std::string get_wireless_backhaul_mac() override;
 
+    virtual bool enable_disable_ep(bool enable) override;
+
     virtual bool process_dwpal_event(char *ifname, char *buffer, int bufLen,
                                      const std::string &opcode) override;
     virtual bool process_dwpal_nl_event(struct nl_msg *msg, void *arg = nullptr) override;
diff --git a/common/beerocks/bwl/include/bwl/sta_wlan_hal.h b/common/beerocks/bwl/include/bwl/sta_wlan_hal.h
index 14564ba3a..27a075834 100644
--- a/common/beerocks/bwl/include/bwl/sta_wlan_hal.h
+++ b/common/beerocks/bwl/include/bwl/sta_wlan_hal.h
@@ -94,6 +94,8 @@ public:
      * Returns the Backhaul Mac address.
      */
     virtual std::string get_wireless_backhaul_mac() = 0;
+
+    virtual bool enable_disable_ep(bool enable) = 0;
 };
 
 // STA HAL factory types
diff --git a/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.cpp
index af9e7fb40..a263c26f9 100644
--- a/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.cpp
@@ -285,6 +285,8 @@ std::string sta_wlan_hal_nl80211::get_wireless_backhaul_mac()
     return mac;
 }
 
+bool sta_wlan_hal_nl80211::enable_disable_ep(bool enable) { return true; }
+
 bool sta_wlan_hal_nl80211::process_nl80211_event(parsed_obj_map_t &parsed_obj)
 {
     auto opcode = parsed_obj["_opcode"];
diff --git a/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.h b/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.h
index 14650f9fa..cc82cf605 100644
--- a/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.h
+++ b/common/beerocks/bwl/nl80211/sta_wlan_hal_nl80211.h
@@ -69,6 +69,8 @@ public:
     std::string get_bssid() override;
     std::string get_wireless_backhaul_mac() override;
 
+    virtual bool enable_disable_ep(bool enable) override;
+
 protected:
     virtual bool process_nl80211_event(parsed_obj_map_t &parsed_obj) override;
 
diff --git a/common/beerocks/bwl/whm/sta_wlan_hal_whm.cpp b/common/beerocks/bwl/whm/sta_wlan_hal_whm.cpp
index 149dd38ff..d7756f77a 100644
--- a/common/beerocks/bwl/whm/sta_wlan_hal_whm.cpp
+++ b/common/beerocks/bwl/whm/sta_wlan_hal_whm.cpp
@@ -614,6 +614,12 @@ std::string sta_wlan_hal_whm::get_wireless_backhaul_mac()
     return mac;
 }
 
+bool sta_wlan_hal_whm::enable_disable_ep(bool enable)
+{
+    LOG(TRACE) << __func__ << " - NOT IMPLEMENTED";
+    return true;
+}
+
 bool sta_wlan_hal_whm::update_status()
 {
     Endpoint endpoint;
diff --git a/common/beerocks/bwl/whm/sta_wlan_hal_whm.h b/common/beerocks/bwl/whm/sta_wlan_hal_whm.h
index caba815d8..3f8e4424b 100644
--- a/common/beerocks/bwl/whm/sta_wlan_hal_whm.h
+++ b/common/beerocks/bwl/whm/sta_wlan_hal_whm.h
@@ -67,6 +67,8 @@ public:
     std::string get_bssid() override;
     std::string get_wireless_backhaul_mac() override;
 
+    virtual bool enable_disable_ep(bool enable) override;
+
     const std::vector<int> &get_ext_events_fds() const;
     bool unique_file_descriptors() const { return false; }
 
diff --git a/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_action.h b/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_action.h
index 38068bf49..7279b7e91 100644
--- a/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_action.h
+++ b/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_action.h
@@ -145,7 +145,8 @@ enum eActionOp_BACKHAUL: uint8_t {
     ACTION_BACKHAUL_RECONNECT_COMMAND = 0x51,
     ACTION_BACKHAUL_TRIGGER_ON_BOOT_SCAN = 0x52,
     ACTION_BACKHAUL_WIFI_CREDENTIALS_UPDATE_REQUEST = 0x53,
-    ACTION_BACKHAUL_ENUM_END = 0x54,
+    ACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT = 0x54,
+    ACTION_BACKHAUL_ENUM_END = 0x55,
 };
 
 enum eActionOp_PLATFORM: uint8_t {
diff --git a/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_backhaul.h b/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_backhaul.h
index 0836fe008..695b4d8a2 100644
--- a/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_backhaul.h
+++ b/common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_backhaul.h
@@ -738,6 +738,29 @@ class cACTION_BACKHAUL_WIFI_CREDENTIALS_UPDATE_REQUEST : public BaseClass
         int m_lock_order_counter__ = 0;
 };
 
+class cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT : public BaseClass
+{
+    public:
+        cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT(uint8_t* buff, size_t buff_len, bool parse = false);
+        explicit cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT(std::shared_ptr<BaseClass> base, bool parse = false);
+        ~cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT();
+
+        static eActionOp_BACKHAUL get_action_op(){
+            return (eActionOp_BACKHAUL)(ACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT);
+        }
+        sMacAddr& radio_mac();
+        uint8_t& enable();
+        void class_swap() override;
+        bool finalize() override;
+        static size_t get_initial_size();
+
+    private:
+        bool init();
+        eActionOp_BACKHAUL* m_action_op = nullptr;
+        sMacAddr* m_radio_mac = nullptr;
+        uint8_t* m_enable = nullptr;
+};
+
 class cACTION_BACKHAUL_TRIGGER_ON_BOOT_SCAN : public BaseClass
 {
     public:
diff --git a/common/beerocks/tlvf/AutoGenerated/src/beerocks/tlvf/beerocks_message_backhaul.cpp b/common/beerocks/tlvf/AutoGenerated/src/beerocks/tlvf/beerocks_message_backhaul.cpp
index 2ac04e3e9..a164fe481 100644
--- a/common/beerocks/tlvf/AutoGenerated/src/beerocks/tlvf/beerocks_message_backhaul.cpp
+++ b/common/beerocks/tlvf/AutoGenerated/src/beerocks/tlvf/beerocks_message_backhaul.cpp
@@ -2464,6 +2464,86 @@ bool cACTION_BACKHAUL_WIFI_CREDENTIALS_UPDATE_REQUEST::init()
     return true;
 }
 
+cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT::cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT(uint8_t* buff, size_t buff_len, bool parse) :
+    BaseClass(buff, buff_len, parse) {
+    m_init_succeeded = init();
+}
+cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT::cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT(std::shared_ptr<BaseClass> base, bool parse) :
+BaseClass(base->getBuffPtr(), base->getBuffRemainingBytes(), parse){
+    m_init_succeeded = init();
+}
+cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT::~cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT() {
+}
+sMacAddr& cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT::radio_mac() {
+    return (sMacAddr&)(*m_radio_mac);
+}
+
+uint8_t& cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT::enable() {
+    return (uint8_t&)(*m_enable);
+}
+
+void cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT::class_swap()
+{
+    tlvf_swap(8*sizeof(eActionOp_BACKHAUL), reinterpret_cast<uint8_t*>(m_action_op));
+    m_radio_mac->struct_swap();
+}
+
+bool cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT::finalize()
+{
+    if (m_parse__) {
+        TLVF_LOG(DEBUG) << "finalize() called but m_parse__ is set";
+        return true;
+    }
+    if (m_finalized__) {
+        TLVF_LOG(DEBUG) << "finalize() called for already finalized class";
+        return true;
+    }
+    if (!isPostInitSucceeded()) {
+        TLVF_LOG(ERROR) << "post init check failed";
+        return false;
+    }
+    if (m_inner__) {
+        if (!m_inner__->finalize()) {
+            TLVF_LOG(ERROR) << "m_inner__->finalize() failed";
+            return false;
+        }
+        auto tailroom = m_inner__->getMessageBuffLength() - m_inner__->getMessageLength();
+        m_buff_ptr__ -= tailroom;
+    }
+    class_swap();
+    m_finalized__ = true;
+    return true;
+}
+
+size_t cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT::get_initial_size()
+{
+    size_t class_size = 0;
+    class_size += sizeof(sMacAddr); // radio_mac
+    class_size += sizeof(uint8_t); // enable
+    return class_size;
+}
+
+bool cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT::init()
+{
+    if (getBuffRemainingBytes() < get_initial_size()) {
+        TLVF_LOG(ERROR) << "Not enough available space on buffer. Class init failed";
+        return false;
+    }
+    m_radio_mac = reinterpret_cast<sMacAddr*>(m_buff_ptr__);
+    if (!buffPtrIncrementSafe(sizeof(sMacAddr))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(sMacAddr) << ") Failed!";
+        return false;
+    }
+    if (!m_parse__) { m_radio_mac->struct_init(); }
+    m_enable = reinterpret_cast<uint8_t*>(m_buff_ptr__);
+    if (!buffPtrIncrementSafe(sizeof(uint8_t))) {
+        LOG(ERROR) << "buffPtrIncrementSafe(" << std::dec << sizeof(uint8_t) << ") Failed!";
+        return false;
+    }
+    if (m_parse__) { class_swap(); }
+    return true;
+}
+
 cACTION_BACKHAUL_TRIGGER_ON_BOOT_SCAN::cACTION_BACKHAUL_TRIGGER_ON_BOOT_SCAN(uint8_t* buff, size_t buff_len, bool parse) :
     BaseClass(buff, buff_len, parse) {
     m_init_succeeded = init();
diff --git a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml
index 9341a2a90..445d00943 100755
--- a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml
+++ b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml
@@ -148,8 +148,9 @@ eActionOp_BACKHAUL:
   ACTION_BACKHAUL_RECONNECT_COMMAND: 81
   ACTION_BACKHAUL_TRIGGER_ON_BOOT_SCAN: 82
   ACTION_BACKHAUL_WIFI_CREDENTIALS_UPDATE_REQUEST: 83
+  ACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT: 84
 
-  ACTION_BACKHAUL_ENUM_END: 84
+  ACTION_BACKHAUL_ENUM_END: 85
 
 ####################################################
 ####################################################
diff --git a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
index a1ab36798..7a422655d 100755
--- a/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
+++ b/common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
@@ -161,6 +161,12 @@ cACTION_BACKHAUL_WIFI_CREDENTIALS_UPDATE_REQUEST:
   wifi_credentials:
     _type: WSC::cConfigData
 
+cACTION_BACKHAUL_WIFI_ENABLE_DISABLE_ENDPOINT:
+  _type: class
+  radio_mac:
+    _type: sMacAddr
+  enable : uint8_t #bool
+
 ################################################
 # CHANNEL SCAN REQUESTS/RESPONSES
 ################################################
-- 
2.34.1

